### PR TITLE
[SPARK-40943][SQL] Make `MSCK` keyword optional in `REPAIR TABLE` syntax

### PR DESCRIPTION
**Pull request by ben-zhang**
### What changes were proposed in this pull request?
Make the `MSCK` keyword optional in `MSCK REPAIR TABLE` commands so that it can be omitted.

### Why are the changes needed?
The use of the keyword `MSCK`, meaning metastore check, is arcane and does not add value for the command. Removing it makes the meaning of `REPAIR TABLE` commands more clear.

The [Spark documentation page](https://spark.apache.org/docs/latest/sql-ref-syntax-ddl-repair-table.html) for this command is titled "REPAIR TABLE". It only mentions the MSCK keyword in the content of the page, indicating that this keyword is not necessary for the semantic understanding of the command.

Additionally, here is a reference from [MySQL](https://dev.mysql.com/doc/refman/8.0/en/repair-table.html), which completely omits the MSCK keyword.

### Does this PR introduce _any_ user-facing change?

Yes, previously, it was no possible to specify only `REPAIR TABLE` without `MSCK`. Now, it is possible. No changes to existing behaviour using the original `MSCK REPAIR TABLE` syntax are in this PR.

### How was this patch tested?
Unit tests.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Could you provide other DBMS references for `REPAIR TABLE` syntax without `MSCK`, @ben-zhang ?

**Comment by bjornjorgensen** [comment]
Hi, can you enable GitHub Actions on yours repo?

**Comment by ben-zhang** [comment]
> Could you provide other DBMS references for `REPAIR TABLE` syntax without `MSCK`, @ben-zhang ?

Hey @dongjoon-hyun, thanks for asking. 
Here's a [reference from MYSQL](https://dev.mysql.com/doc/refman/8.0/en/repair-table.html).

Also, the Spark documentation page for this command is titled "REPAIR TABLE". It only mentions the MSCK keyword in the content of the page. To me, this is an acknowledgement that the meaning of the command is tied to just "REPAIR TABLE".

**Comment by MaxGekk** [comment]
> Also, the [Spark documentation page](https://spark.apache.org/docs/latest/sql-ref-syntax-ddl-repair-table.html) for this command ...

@ben-zhang Could you update the doc, and mark the `MSCK` keyword as optional one:
Syntax
```
[MSCK] REPAIR TABLE table_identifier [{ADD|DROP|SYNC} PARTITIONS]
```

**Comment by ben-zhang** [comment]
@MaxGekk Updated.

**Review by dongjoon-hyun — DISMISSED** [review]
If MySQL is the only DBMS for this syntax, I'm negative about this change because this PR only increases the chance of SQL incompatibility. Is there any other support cases?

**Comment by dongjoon-hyun** [comment]
cc @sunchao since he is the Hive PMC member and Spark committer.

**Comment by srielau** [comment]
> If MySQL is the only DBMS for this syntax, I'm negative about this change because this PR only increases the chance of SQL incompatibility. Is there any other support cases?

IMHO mySQL was onto something to not include MSCK (who is the chicken and who the egg here anyway?).
It seems like everyone (e.g. IBM) just copied Hive to not rock the boat.

Interestingly our own docs omit MSCK from the title (and the file name):
https://spark.apache.org/docs/3.0.0-preview/sql-ref-syntax-ddl-repair-table.html

It seems the only purpose of this keyword is to trip people up.
As perhaps the most popular platform, we have the chance to lead here and make things easier.

**Comment by dongjoon-hyun** [comment]
May I ask if dropping `MSCK` is the official Databricks's direction in order to do the lead, @srielau ?
> As perhaps the most popular platform, we have the chance to lead here and make things easier.

**Comment by srielau** [comment]
> May I ask if dropping `MSCK` is the official Databricks's direction in order to do the lead, @srielau ?
> 
> > As perhaps the most popular platform, we have the chance to lead here and make things easier.

Yes, Databricks will make MSCK a noise word. I will probably remove it from our documented syntax and add a footnote such as "for compatibility with Hive you may also use MSCK REPAIR TABLE".

**Comment by cloud-fan** [comment]
Making `MSCK` optional makes our SQL syntax more flexible (and simpler), and compatible with MySQL. And there seems no down side. @dongjoon-hyun Can you share your concerns? We can discuss more here.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 1]
```diff
@@ -1728,6 +1728,14 @@ class DataSourceV2SQLSuiteV1Filter extends DataSourceV2SQLSuite with AlterTableT
     }
   }
 
+  test("REPAIR TABLE omit MSCK") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:1731]: can we move the test to `MsckRepairTableParserSuite`? And probably rename the test suite. And also `MsckRepairTableSuiteBase` and sub classes.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:1731]: oh this is not parser test. But I think a parser change only needs parser test, no end-to-end test.

Comment (ben-zhang) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:1731]: Sounds good, I can remove this one. I agree that e2e tests are not necessary for a parser change.

**Review by dongjoon-hyun — COMMENTED** [review]
@cloud-fan . I'm not sure this is a feature has practical merits. 

- This is the opposite direction from [Hive](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-RecoverPartitions(MSCKREPAIRTABLE)) because Hive can omit `REPAIR` like `MSCK [REPAIR] TABLE`.
- [Athena](https://docs.aws.amazon.com/athena/latest/ug/msck-repair-table.html) doesn't support omit.
- Both [Databricks AWS](https://docs.databricks.com/sql/language-manual/sql-ref-syntax-ddl-repair-table.html) and [Databricks Azure](https://learn.microsoft.com/en-us/azure/databricks/spark/latest/spark-sql/language-manual/sql-ref-syntax-ddl-repair-table) seem to not work in that way yet.

If MySQL is the only one in SQL ecosystem, I believe we should be the last one for this direction. Apache Spark community knows the pain during the SQL engine migrations and this seems to lead us to the foreseeable troublesome future. And, technically, this looks like a kind of `Lock-In` feature for Spark SQL because this will block a migration from Spark SQL to other SQL Engines.

I hope we can hold on this PR and discuss this later after Databricks actually documents it in the official website like https://github.com/apache/spark/pull/38433#issuecomment-1297817456.

**Comment by ben-zhang** [comment]
@dongjoon-hyun , sounds good. I will get back to you when the Databricks docs are updated.

**Comment by ben-zhang** [comment]
Hey @dongjoon-hyun, you can see the updated docks here: https://docs.databricks.com/sql/language-manual/sql-ref-syntax-ddl-repair-table.html

Shall we revisit this idea?

**Comment by dongjoon-hyun** [comment]
Thank you for updates, @ben-zhang . +1 for resuming this PR. Could you rebase this PR, @ben-zhang ?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. (Pending CIs).

**Comment by dongjoon-hyun** [comment]
Also cc @bjornjorgensen , @cloud-fan , @MaxGekk , and @xinrong-meng for Aapche Spark 3.4 release manager.

**Comment by ben-zhang** [comment]
Thanks for being so prompt on the revival of this PR, @dongjoon-hyun. Appreciate it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/MsckRepairTableParserSuite.scala [hunk 2]
```diff
@@ -31,6 +31,15 @@ class MsckRepairTableParserSuite extends AnalysisTest {
         enableDropPartitions = false))
   }
 
+  test("repair a table without MSCK keyword") {
+    comparePlans(
+      parsePlan("REPAIR TABLE a.b.c"),
+      RepairTable(
+        UnresolvedTable(Seq("a", "b", "c"), "MSCK REPAIR TABLE", None),
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/MsckRepairTableParserSuite.scala:38]: Shouldn't we change the hint if `REPAIR TABLE` becomes the main way of invoking the command?

**Comment by bjornjorgensen** [comment]
Yes, I think this is ok. Now that Databrick has it, I hope spark shall also have it the same way.

**Comment by dongjoon-hyun** [comment]
@ben-zhang . Could you address @MaxGekk 's review comment?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/MsckRepairTableParserSuite.scala (hunk 2 above)
Comment (ben-zhang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/MsckRepairTableParserSuite.scala:38]: Agreed, this should be updated. Will do this now and then push again.

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/MsckRepairTableParserSuite.scala:38]:
```suggestion
        UnresolvedTable(Seq("a", "b", "c"), "REPAIR TABLE", None),
```

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/MsckRepairTableParserSuite.scala:38]: According to the agreement here, I'll merge @MaxGekk 's suggestion and backport to branch-3.4 for Apache Spark 3.4 release.

**Review by dongjoon-hyun — DISMISSED** [review]

**Comment by dongjoon-hyun** [comment]
Oh, the last commit seems to cause a test failure.

**Comment by MaxGekk** [comment]
@dongjoon-hyun Need to change this one too:
```scala
    RepairTable(
      createUnresolvedTable(ctx.multipartIdentifier, s"MSCK REPAIR TABLE$option"),
```

**Comment by dongjoon-hyun** [comment]
If we touch `AstBuilder.scala`, it could cause more failures, couldn't it?

**Comment by MaxGekk** [comment]
> If we touch AstBuilder.scala, it could cause more failures, couldn't it?

Highly likely, all parse/error tests related to `REPAIR TABLE`.

**Comment by dongjoon-hyun** [comment]
Sorry, but let's revert your last commit back and merge this PR without addressing your comments. Is it okay, @MaxGekk ?

**Comment by MaxGekk** [comment]
> Sorry, but let's revert your last commit back and merge this PR without addressing your comments. Is it okay, @MaxGekk ?

I just don't like the consistency between the docs where we talk only about `REPAIR TABLE` and the error message that will seems outdated after the PR. But we can make the changes later, for sure.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. I'll fix the authorship to a single authorship of @ben-zhang during merging.

I verified the changed part due to the last too commits.
```
[info] MsckRepairTableParserSuite:
[info] - repair a table (226 milliseconds)
[info] - repair a table without MSCK keyword (2 milliseconds)
[info] - add partitions (2 milliseconds)
[info] - drop partitions (1 millisecond)
[info] - sync partitions (2 milliseconds)
[info] Run completed in 964 milliseconds.
[info] Total number of tests run: 5
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 5, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 33 s, completed Feb 15, 2023, 11:00:14 PM
```

**Comment by ben-zhang** [comment]
@MaxGekk @dongjoon-hyun agreed, the inconsistency with error messages is strange. I will create a follow-up for that.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.4 for Apache Spark 3.4.

Thank you, @ben-zhang , @cloud-fan , @MaxGekk , @bjornjorgensen , @srielau .

Also, I added you to the Apache Spark contributor group and assigned SPARK-40943 to you, @ben-zhang .
Welcome to the Apache Spark community (OSS).

**Pull request closed without merging**